core/templates/domain/skill/concept-card-backend-api.service.ts

```diff
@@ -17,92 +17,123 @@
  * about the concept card of a skill from the backend.
  */
 
-require('domain/utilities/url-interpolation.service.ts');
-require('domain/skill/skill-domain.constants.ajs.ts');
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
 
-angular.module('oppia').factory('ConceptCardBackendApiService', [
-  '$http', '$q', 'UrlInterpolationService', 'CONCEPT_CARD_DATA_URL_TEMPLATE',
-  function($http, $q, UrlInterpolationService, CONCEPT_CARD_DATA_URL_TEMPLATE) {
-    // Maps previously loaded concept cards to their IDs.
-    var _conceptCardCache = [];
+import cloneDeep from 'lodash/cloneDeep';
 
-    var _fetchConceptCards = function(
-        skillIds, successCallback, errorCallback) {
-      var conceptCardDataUrl = UrlInterpolationService.interpolateUrl(
-        CONCEPT_CARD_DATA_URL_TEMPLATE, {
-          comma_separated_skill_ids: skillIds.join(',')
-        });
+import { ConceptCardObjectFactory } from
+  'domain/skill/ConceptCardObjectFactory';
+import { RecordedVoiceoversObjectFactory } from
+  'domain/exploration/RecordedVoiceoversObjectFactory';
+import { SkillDomainConstants } from
+  'domain/skill/skill-domain.constants';
+import { SubtitledHtmlObjectFactory } from
+  'domain/exploration/SubtitledHtmlObjectFactory';
+import { UrlInterpolationService } from
+  'domain/utilities/url-interpolation.service';
+import { VoiceoverObjectFactory } from
+  'domain/exploration/VoiceoverObjectFactory';
+import { WorkedExampleObjectFactory } from
+  'domain/skill/WorkedExampleObjectFactory';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ConceptCardBackendApiService {
+  constructor(
+    private http: HttpClient,
+    private urlInterpolation: UrlInterpolationService) {}
 
-      $http.get(conceptCardDataUrl).then(function(response) {
-        var conceptCards = angular.copy(response.data.concept_card_dicts);
+  // Maps previously loaded concept cards to their IDs.
+  private _conceptCardCache = [];
+  private conceptCardObjectFactory = null;
+
+  private _fetchConceptCards(
+      skillIds: Array<string>,
+      successCallback: (value?: Object | PromiseLike<Object>) => void,
+      errorCallback: (reason?: string) => void): void {
+    var conceptCardDataUrl = this.urlInterpolation.interpolateUrl(
+      SkillDomainConstants.CONCEPT_CARD_DATA_URL_TEMPLATE, {
+        comma_separated_skill_ids: skillIds.join(',')
+      });
+      
+    var conceptCardObjects = [];
+    var conceptCardObjectFactory = new ConceptCardObjectFactory(
+      new SubtitledHtmlObjectFactory(),
+      new RecordedVoiceoversObjectFactory(new VoiceoverObjectFactory()),
+      new WorkedExampleObjectFactory(new SubtitledHtmlObjectFactory()));
+
+    this.http.get(conceptCardDataUrl).toPromise().then(
+      (response: any) => {
```

Comment:
Please define a type for `response`.

Reply:
I'm not sure what type to use for `HttpResponse`. The `concept_card_dicts` field of the response is an array of `IConceptCardBackendDict`. Does that mean I should use `HttpResponse<IConceptCardBackendDict[]>` for the type of the response?

Reply:
It should be something like 
```typescript
this.http.get<IConceptCardBackendDict[]>(conceptCardDataUrl).toPromise().then(response => {...});
```

Reply:
I currently did it like this, but when I run the typescript checks I get this error: `core/templates/domain/skill/concept-card-backend-api.service.ts(67,43): error TS2339: Property 'concept_card_dicts' does not exist on type 'IConceptCardBackendDict[]'.` I think it's because the response is an object whose property is `IConceptCardBackendDict[]` rather than just an array of concept card dicts? I'm not sure how to represent that as a type though.

Reply:
Oh, ok. In that case please declare an interface something like
```typescript
interface IConceptCardsBackendDict {
  'concept_card_dicts': IConceptCardBackendDict[];
}
```
and then you could do
```typescript
this.http.get<IConceptCardsBackendDict>(conceptCardDataUrl).toPromise().then(response => {...});
```

Reply:
Thank you, fixed!